bindings/cpp/test_array.cpp

```diff
@@ -0,0 +1,115 @@
+#include "array.h"
+#include "datapipeline.hh"
+
+#include <iostream>
+
+#include "pybind11/embed.h"
+#include "pybind11/numpy.h"
+#include "pybind11/stl.h"
+namespace py = pybind11;
+using namespace pybind11::literals;
+
+const std::string TEST_HDF5_FILENAME = "test_npy.h5";
+const char *TEST_DATASET_NAME = "nparray";
+
+/// std api applies here, but python open(hdf5 file is not impl, use h5py to open file
+void write_array(const string &data_product, const string &component, const Array &da)
+{
+  py::module h5py = py::module::import("h5py");
+  const py::array pya = da.encode();
+  //py::print(pya);
+  auto h5file = h5py.attr("File")(data_product, "w");
+
+  py::list shape = py::cast(da.shape());
+  //auto tk = pya.dtype().kind(); // maybe must be a py::arg()
+  //py::object dataset = h5file.attr("create_dataset")("nparray", shape, pya.dtype());
+  py::object dataset = h5file.attr("create_dataset")(component, "data"_a = pya);
+  //
+  dataset.attr("write_direct")(pya);
+
+  py::object attrs = dataset.attr("attrs");
+  attrs.attr("__setitem__")(py::str("unit"), da.unit());
+  //  attrs.attr("__setitem__")(py::str("title"), da.title());
+  auto dims = ((ArrayT<int64_t> &)da).dims();
+  for (size_t i = 0; i < dims.size(); i++)
+  {
+    std::string dn = "dim_" + std::to_string(i);
+    py::array _dv = py::cast(dims[i].values);
+    attrs.attr("__setitem__")(py::str(dn + "_values"), _dv);
+    attrs.attr("__setitem__")(py::str(dn + "_unit"), py::cast(dims[i].unit));
+    py::str dtitle = py::cast(dims[i].title);
+    attrs.attr("__setitem__")(py::str(dn + "_title"), dtitle);
+  }
+  // no need to close dataset?
+  h5file.attr("close")();
+}
+
+template <typename DT>
+typename ArrayT<DT>::Ptr read_array(const string &data_product, const string &component)
```

Comment:
Same comment as write_array; this work is not needed.

Reply:
will be remove once, rebase Chris latest code